### PR TITLE
Mifare Classic gen1/gen2 (+ regular) card writing support

### PR DESCRIPTION
**Pull request by Foxushka**
Release 1.0 coming right after merging this

### CODE REVIEW

**Comment by GameTec-live** [comment]
Selecting "Skip default dictionary" Followed by "Empty 0 Keys" and then trying recover keys gives a index out of range in typed data patch... 
image

**Comment by GameTec-live** [comment]
Trying to save only the uid of a non mifare classic card (was possible before) causes a null exceptions in saveHFCard
image

**Comment by GameTec-live** [comment]
No real error, but trying to write a mifare save to a non mifare card just stalls? maybe displaying a warning message or smthng?

**Comment by GameTec-live** [comment]
oh yeah, step 3 also needs to be scrollable or else it gets cut off depending on window size...

**Comment by GameTec-live** [comment]
and writing to non magic cards always gives a "smthng went wrong"... (prob because of block 0 verification)
Id say, as discussed, when smthng fails to write, display the blocks that failed to write and then let the user choose if its a failure... (And maybe if only block 0 failed to write display a info banner that this is normal behaviour for non magic cards)

**Comment by GameTec-live** [comment]
writing a uid only card gives a invalid value not inclusive range...
image

**Comment by GameTec-live** [comment]
writing to gen2/generic is broken in general? got out my 1k magic gen 2, dumped a 1k (not 1k ev1, etc, just 1k), tried writing and theres always a error in card writing

**Comment by GameTec-live** [comment]
oh yeah and writing needs to give a popup on the lite (just like when reading) or else everything breaks...

**Comment by GameTec-live** [comment]
Have a log
[log.log](https://github.com/GameTec-live/ChameleonUltraGUI/files/13693901/log.log)

File: chameleonultragui/lib/helpers/mifare_classic/write/gen1.dart [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  MifareClassicGen1WriteHelper(super.communicator, {required super.recovery});
+
+  @override
+  String get name => "Gen1";
+
+  @override
+  Future<bool> isMagic(dynamic data) async {
+    await communicator.send14ARaw(Uint8List(1)); // reset
+
+    Uint8List data = await communicator.send14ARaw(Uint8List.fromList([0x40]),
```
Comment (GameTec-live) [inline chameleonultragui/lib/helpers/mifare_classic/write/gen1.dart:15]: why do it via raw and not the predefined commands? (if there are any)

**Comment by GameTec-live** [comment]
oh yeah, just noticed, card export is borked too?

**Comment by LupusE** [comment]
Worked fine.
- Loaded the latest .apk build
- Dumped a Mifare Classic 1k with Proxmark3
- Opened CU GUI -> write, selected the Dump (.json)
- Autodetected Gen1A 1K S50
- Write dump
- Verified with Flipper Zero

**Comment by VarChar42** [comment]
Works on windows using Mifare Classic 1K Gen1 EV.
Cloned it to Lab401 Mifare Classic 1K.

But only with non corrupted save file (#400).

File: chameleonultragui/lib/gui/menu/slot_export.dart [hunk 2]
```diff
@@ -24,11 +25,10 @@ class SlotExportMenu extends StatefulWidget {
   final SlotTypes slotTypes;
 
   const SlotExportMenu(
-      {Key? key,
+      {super.key,
       required this.names,
       required this.enabledSlotInfo,
-      required this.slotTypes})
-      : super(key: key);
+      required this.slotTypes});
```
Comment (GameTec-live) [inline chameleonultragui/lib/gui/menu/slot_export.dart:31]: id leave this as be until a proper solution is found...

**Comment by GameTec-live** [comment]
image
Running autodetect on a T55 for example throws a invalid data length... Other than that, everything seems to work fine?

**Review by GameTec-live — APPROVED** [review]
Works

**Comment by unknown10777** [comment]
On iPhone SE(2020) writing card cannot display normally, some are out of screen.

**Comment by GameTec-live** [comment]
> On iPhone SE(2020) writing card cannot display normally, some are out of screen.

aw, right, sht... That entire thing needs to be a scrollview... ok, next hotfix or pr ig... @Foxushka

**Comment by unknown10777** [comment]
It's really perfect though, thanks for update!

**Comment by unknown10777** [comment]
image
And tried to write card have unknown keys, got keys using nested, and the button is blank. Is it expected? @Foxushka

**Pull request merged**